### DISCUSSION TITLE
Form valdiation using Zod

### DISCUSSION THREAD

**Discussion by Kay2dan** [react-hook-form/react-hook-form#8091] [Q&A]
Hi guys, I reviewed the issues & even though people have been clarified on how to get `formState.isValid` property to work correctly, I am struggling to have it return true, even when the form is valid. My form is a bit complex and there is a lot of code, which I tried to simplify in the codesandbox example below. However, I would greatly appreciate what I am doing wrong.

The schema & its integration is as such:

````javascript
// For the form to be valid:
// Atleast 2 goals, with min. 1 task each
export const schemaZod = z.object({
  goals: z
    .object({
      title: z.string().min(3, { message: "required" }).max(40),
      intro: z.string().max(250).optional(),
      deadline: (z.string() || z.date()).optional(),
      task: z
        .object({
          content: z.string().min(3, { message: "required" }).max(50),
          dayFrequency: z.number().min(0).max(5).optional(),
          weekFrequency: z.number().min(0).max(7).optional(),
        })
        .array()
        .min(1),
    })
    .array()
    .min(2),
});

  const ctx = useForm({
    mode: "onChange",
    resolver: zodResolver(schemaZod),
  });
````

Here is a working example on [Codesandox](https://codesandbox.io/s/reacthookform-validation-dj26qj?file=/src/Tasks/GoalTaskForm.js).

[Stackoverflow question](https://stackoverflow.com/questions/71604032/react-hook-form-validation-isvalid-prop-keeps-returning-false).

**Comment by Moshyfawn** [comment]
Hey, @Kay2dan!

I believe, your validation schema has a slight issue. Zod has [the `.or()` method](https://github.com/colinhacks/zod#or) you want to use instead of the logic `OR` operator
```diff
const schemaZod = z.object({
  goals: z
    .object({
      title: z.string().min(3, { message: "required" }).max(40),
      intro: z.string().max(250).optional(),
-     deadline: (z.string() || z.date()).optional(),
+     deadline: z.string().or(z.date()).optional(),
      task: z
        .object({
          content: z.string().min(3, { message: "required" }).max(50),
          dayFrequency: z.number().min(0).max(5).optional(),
          weekFrequency: z.number().min(0).max(7).optional()
        })
        .array()
        .min(1)
    })
    .array()
});
```

**Reply by Kay2dan** [reply]
Thank you, I have updated the code, however, that does not solve the schema. One thing I have noticed is that `.array().min(2)` does not work correctly.

**Reply by Moshyfawn** [reply]
So, I've created [a simplified CSB](https://codesandbox.io/s/reacthookform-validation-forked-od7yyi?file=/src/Form.tsx) that, I believe, handles the `isValid` state properly. Please, take a look and see if it complies with your requirements.

@bluebill1049, can you take a look at [the provided CSB](https://codesandbox.io/s/reacthookform-validation-forked-od7yyi?file=/src/Form.tsx). It works fine in that the `isValid` state is correct _(I believe)_. But the field array error object doesn't.. update?.. properly?

According to the Zod validation schema, there should be at least 3 goals, which works fine, but the error message isn't cleared on requirements satisfaction. I know there's a TS issue with have the field array errors _(not it's fields, but the actual array field)_, but worked fine logic-wise

**Reply by bluebill1049** [reply]
hook form focus on field level error and update render according to the associated field name, probably an issue just need a `trigger()` if you want to get all the errors, and yes we do have limitation with object vs array which i am planning to resolve at v8 since errors object is fixed in v8.

**Reply by bluebill1049** [reply]
@Kay2dan  an easy debug an issue with schema validation is pretty simple, you can just need the following

```tsx
resolver: (data, context, options) => {
  console.log(data)
  // you can debug your validation schema here
  return zodResolver(schema)(data, context, options)
},
```

**Reply by Kay2dan** [reply]
Thank you Moshyfawn & BlueBill for looking into this. I am working on another feature right now but will review this (its priority to get validation working correctly) within the weekend.

**Reply by Moshyfawn** [reply]
No problem!
Great coding to ya! ;)

**Reply by Kay2dan** [reply]
@Moshyfawn Thank you for taking the time to create the mockup. That was immensely helpful and it seems, my zod schema was incorrect. 

@bluebill1049  thank you for the tip & RHF Bill. Great stuff.

**Reply by Moshyfawn** [reply]
I'm glad I could help!

**Reply by bluebill1049** [reply]
I am including the above information in the doc, so it will help others to debug as well. 🙏

**Comment by alifa20** [comment]
@Kay2dan  I guess this is the link to the stackoverflow: https://stackoverflow.com/questions/71604032/react-hook-form-validation-isvalid-prop-keeps-returning-false 
?
